openedx/openedx-platform
# [FAL-2457] Let ModuleSystem use UserService

## Description

Re-submits the changes from https://github.com/edx/edx-platform/pull/28440 which were reverted by https://github.com/edx/edx-platform/pull/28857, and fixes the bug that caused this revert.

Initial step towards combining the ModuleSystem and DescriptorSystem classes.
Adds a shim which deprecates these ModuleSystem attributes, pulling them from the `user` service instead of requiring them in the constructor.

* `anonymous_student_id`
* `seed`
* `user_id`
* `user_is_staff`

This change is only a refactoring, and should not affect Learners, Course Authors, or anyone else using edx-platform.

## Supporting information

* [Content Core Platform Simplication project](https://openedx.atlassian.net/wiki/spaces/AC/pages/3091629503/Content+Core+Platform+Simplification)
* [BLENDED-113](https://openedx.atlassian.net/browse/BLENDED-113) (edX internal ticket)
* [FAL-2247](https://tasks.opencraft.com/browse/FAL-2247) (OpenCraft internal ticket)

## Testing instructions

LMS: https://pr29190.sandbox.opencraft.hosting
1. Login as a demo learner (`honor@example.com` / `edx`)
2. Browse around the Demo course and ensure that blocks render as expected, and that you can submit answers to problems.
3. Switch to the Legacy Course Experience and browse around in that too.

Studio: https://studio.pr29190.sandbox.opencraft.hosting
1. Login as a demo staff (`staff@example.com` / `edx`)
2. Navigate to the Demo course and browse through the blocks. Ensure they render and edit as expected.
3. Check that you can submit answers to problems, and save changes to blocks.
4. Preview unpublished changes and ensure they render as expected.
4. Publish changes, and ensure they render in the LMS as expected.

Test the anonymous User ID bug:

This bug occurs only in the Legacy Course Experience mode, for Problem blocks which rely on the `anonymous_student_id` context variable.

For OpenCraft staff:
* [Appserver 1](https://manage.opencraft.com/instance/31260/edx-appserver/21221/) contains the broken code, so you can verify the bug there.
* [Appserver 3](https://manage.opencraft.com/instance/31260/edx-appserver/21230/) is running the fixed code on this branch, so you can activate that appserver to verify the bug is fixed.

1. Login to sandbox as a demo staff (`staff@example.com` / `edx`)
2. Visit the Anonymous User ID Test Course and navigate to the [example problem in the Legacy Course Experience](https://pr29190.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+anon+2021_1/courseware/51002af594da44a6a4c0e0c58b149fb4/7472e97419c142b48dc09ede3aedfcea/1?activate_block_id=block-v1%3AOpenCraft%2Banon%2B2021_1%2Btype%40vertical%2Bblock%40d86a0c0817f04ea58fa0535a546b6090).
   This problem is expecting the ProblemBlock's anonymous student ID as an answer.
3. If verifying the bug exists: note that the answer expected when you first view the problem in the Legacy experience changes after you've submitted the problem.

   If verifying that the bug is fixed: ensure that the answer expected remains consistent after submitting the problem.

## Deadline

ASAP please -- the BD-13 project deliverables are in jeopardy because of delays incurred by this issue.
